core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java

```diff
@@ -0,0 +1,322 @@
+            mUnderFileSystemInputStream.read(data, bytesRead, (int) (bytesToRead - bytesRead));
+        if (read == -1) {
+          break;
+        }
+        bytesRead += read;
+      }
+    }
+    mInStreamPos += bytesRead;
+
+    // We should always read the number of bytes as expected since the UFS file length (hence block
+    // size) should be always accurate.
+    Preconditions
+        .checkState(bytesRead == bytesToRead, PreconditionMessage.NOT_ENOUGH_BYTES_READ.toString(),
+            bytesRead, bytesToRead, mBlockMeta.getUnderFileSystemPath());
+    if (mBlockWriter != null && mBlockWriter.getPosition() < mInStreamPos) {
+      Preconditions.checkState(mBlockWriter.getPosition() >= offset);
+      ByteBuffer buffer = ByteBuffer.wrap(data, (int) (mBlockWriter.getPosition() - offset),
+          (int) (mInStreamPos - mBlockWriter.getPosition()));
+      mBlockWriter.append(buffer.duplicate());
+    }
+    return ByteBuffer.wrap(data, 0, bytesRead);
+  }
+
+  /**
+   * This interface is supposed to be used for sequence block reads.
+   *
+   * @param buf the byte buffer
+   * @return the number of bytes read, -1 if it reaches EOF and none was read
+   * @throws IOException if any I/O errors occur when reading the block
+   */
+  @Override
+  public int transferTo(ByteBuf buf) throws IOException {
+    Preconditions.checkState(!mClosed);
+    if (mUnderFileSystemInputStream == null) {
+      return -1;
+    }
+    if (mBlockMeta.getBlockSize() <= mInStreamPos) {
+      return -1;
+    }
+   // Make a copy of the state to keep track of what we have read in this transferTo call.
+    ByteBuf bufCopy = null;
+    if (mBlockWriter != null) {
+      bufCopy = buf.duplicate();
+      bufCopy.readerIndex(bufCopy.writerIndex());
+    }
+    int bytesToRead =
+        (int) Math.min((long) buf.writableBytes(), mBlockMeta.getBlockSize() - mInStreamPos);
+    int bytesRead = buf.writeBytes(mUnderFileSystemInputStream, bytesToRead);
+
+    if (bytesRead <= 0) {
+      return bytesRead;
+    }
+
+    mInStreamPos += bytesRead;
+
+    if (mBlockWriter != null) {
+      bufCopy.writerIndex(buf.writerIndex());
+      while (bufCopy.readableBytes() > 0) {
+        mBlockWriter.transferFrom(bufCopy);
+      }
+    }
+
+    return bytesRead;
+  }
+
+  /**
+   * Closes the block reader. After this, this block reader should not be used anymore.
+   * This is recommended to be called after the client finishes reading the block. It is usually
+   * triggered when the client unlocks the block.
+   *
+   * @throws IOException if any I/O errors occur
+   */
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+
+    try {
+      // This aborts the block if the block is not fully read.
+      updateBlockWriter(mBlockMeta.getBlockSize());
+
+      // We need to check whether the block is cached before closing the block writer.
+      boolean isBlockCached = isBlockCached();
+      Closer closer = Closer.create();
```

Comment:
because we update the input stream and block writer during the read. They can be closed and re-created multiple times. It seems to an overkill to manage the closer state every time one of them is updated.